tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -17,134 +17,267 @@
+		sortKey->ssup_attno = list_nth_oid(sort_col_idx, i);
+
+		/*
+		 * It isn't feasible to perform abbreviated key conversion, since
+		 * tuples are pulled into mergestate's binary heap as needed.  It
+		 * would likely be counter-productive to convert tuples into an
+		 * abbreviated representation as they're pulled up, so opt out of that
+		 * additional optimization entirely.
+		 */
+		sortKey->abbreviate = false;
+
+		PrepareSortSupportFromOrderingOp(list_nth_oid(sort_ops, i), sortKey);
+	}
+
+	chunk_state->sortkeys = sortkeys;
+}
+
 Node *
 decompress_chunk_state_create(CustomScan *cscan)
 {
-	DecompressChunkState *state;
+	DecompressChunkState *chunk_state;
 	List *settings;
 
-	state = (DecompressChunkState *) newNode(sizeof(DecompressChunkState), T_CustomScanState);
+	chunk_state = (DecompressChunkState *) newNode(sizeof(DecompressChunkState), T_CustomScanState);
 
-	state->csstate.methods = &decompress_chunk_state_methods;
+	chunk_state->csstate.methods = &decompress_chunk_state_methods;
 
 	settings = linitial(cscan->custom_private);
-	state->hypertable_id = linitial_int(settings);
-	state->chunk_relid = lsecond_int(settings);
-	state->reverse = lthird_int(settings);
-	state->decompression_map = lsecond(cscan->custom_private);
-	state->is_segmentby_column = lthird(cscan->custom_private);
 
-	return (Node *) state;
+	Assert(list_length(settings) == 4);
+
+	chunk_state->hypertable_id = linitial_int(settings);
+	chunk_state->chunk_relid = lsecond_int(settings);
+	chunk_state->reverse = lthird_int(settings);
+	chunk_state->batch_merge_append = lfourth_int(settings);
+	chunk_state->decompression_map = lsecond(cscan->custom_private);
+	chunk_state->is_segmentby_column = lthird(cscan->custom_private);
+
+	/* Extract sort info */
+	List *sortinfo = lfourth(cscan->custom_private);
+	build_batch_merge_append_sort_info(chunk_state, sortinfo);
+
+	/* Sort keys should only be present when batch_merge_append is used */
+	Assert(chunk_state->batch_merge_append == true || chunk_state->n_sortkeys == 0);
+	Assert(chunk_state->n_sortkeys == 0 || chunk_state->sortkeys != NULL);
+
+	return (Node *) chunk_state;
+}
+
+/*
+ * Create states to hold information for up to n batches
+ */
+static void
+batch_states_create(DecompressChunkState *chunk_state, int nbatches)
+{
+	Assert(nbatches >= 0);
+
+	chunk_state->n_batch_states = nbatches;
+	chunk_state->batch_states = palloc0(sizeof(DecompressBatchState) * nbatches);
+
+	for (int segment = 0; segment < nbatches; segment++)
+	{
+		DecompressBatchState *batch_state = &chunk_state->batch_states[segment];
+		decompress_initialize_batch_state(chunk_state, batch_state);
+	}
+
+	chunk_state->unused_batch_states =
+		bms_add_range(chunk_state->unused_batch_states, 0, nbatches - 1);
 }
 
 /*
- * initialize column state
+ * Enhance the capacity of existing batch states
+ */
+static void
+batch_states_enlarge(DecompressChunkState *chunk_state, int nbatches)
+{
+	Assert(nbatches > chunk_state->n_batch_states);
```

Comment:
I noticed you're enlarging them by a constant number, should we do it in powers of two to avoid quadratic repalloc?

Reply:
I had also thought about it. But since we initialize the expensive data structures lazily and the heap size is also doubled, I have changed this.

Reply:
Let's add a comment about this.

Reply:
As suggested, I added a comment